idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/fir/evaluate/FirCompileTimeConstantEvaluator.kt

```diff
@@ -0,0 +1,193 @@
+                expectedKind.toConstExpression(source, typeAdjustedValue)
+            } else {
+                this
+            }
+        // Lastly, we should preserve the resolved type of the original function call.
+        return expression.apply {
+            replaceTypeRef(expectedType)
+        }
+    }
+
+    private fun <T> ConstantValueKind<T>.toCompileTimeType(): CompileTimeType {
+        return when (this) {
+            ConstantValueKind.Byte -> CompileTimeType.BYTE
+            ConstantValueKind.Short -> CompileTimeType.SHORT
+            ConstantValueKind.Int -> CompileTimeType.INT
+            ConstantValueKind.Long -> CompileTimeType.LONG
+            ConstantValueKind.Double -> CompileTimeType.DOUBLE
+            ConstantValueKind.Float -> CompileTimeType.FLOAT
+            ConstantValueKind.Char -> CompileTimeType.CHAR
+            ConstantValueKind.Boolean -> CompileTimeType.BOOLEAN
+            ConstantValueKind.String -> CompileTimeType.STRING
+
+            else -> CompileTimeType.BYTE
+        }
+    }
+
+    // Unary operators
+    private fun FirConstExpression<*>.evaluate(function: FirSimpleFunction): FirConstExpression<*>? {
+        if (value == null) return null
+        return evalUnaryOp(
+            function.name.asString(),
+            kind.toCompileTimeType(),
+            value!!
+        )?.let {
+            it.toConstantValueKind()?.toConstExpression(source, it)
+        }
+    }
+
+    // Binary operators
+    private fun FirConstExpression<*>.evaluate(
+        function: FirSimpleFunction,
+        other: FirConstExpression<*>
+    ): FirConstExpression<*>? {
+        if (value == null || other.value == null) return null
+        return evalBinaryOp(
+            function.name.asString(),
+            kind.toCompileTimeType(),
+            value!!,
+            other.kind.toCompileTimeType(),
+            other.value!!
+        )?.let {
+            it.toConstantValueKind()?.toConstExpression(source, it)
+        }
+    }
+
+    ////// KINDS
+
+    private fun FirTypeRef.toConstantValueKind(): ConstantValueKind<out Number>? =
+        when (this) {
+            !is FirResolvedTypeRef -> null
+            !is FirImplicitBuiltinTypeRef -> type.toConstantValueKind()
+            is FirImplicitIntTypeRef -> ConstantValueKind.Int
+            else -> null
+        }
+
+    private fun ConeKotlinType.toConstantValueKind(): ConstantValueKind<out Number>? =
+        when (this) {
+            is ConeKotlinErrorType -> null
+            is ConeLookupTagBasedType -> {
+                when (lookupTag.name.asString()) {
+                    "Byte" -> ConstantValueKind.Byte
+                    "Double" -> ConstantValueKind.Double
+                    "Float" -> ConstantValueKind.Float
+                    "Int" -> ConstantValueKind.Int
+                    "Long" -> ConstantValueKind.Long
+                    "Short" -> ConstantValueKind.Short
```

Comment:
Char, String and Boolean types seems to be missing here
Also,  can be extracted to a function.

Reply:
Extracting it to a function sounds good. 👍 

Those missing types were due to the `out Number` restriction I left as a TODO. Just too lazy to dig further about that, and managed to handle all of them.